### PR TITLE
Added codecarbon tracking to Simvue Run

### PR DESCRIPTION
**Pull request by kzscisoft**
Uses [Code Carbon](https://github.com/mlco2/codecarbon) to track emissions during a Simvue run, this can be disabled via the `config` method.

### CODE REVIEW

**Review by wk9874 — CHANGES_REQUESTED** [review]

File: simvue/eco.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    def out(self, total: "EmissionsData", delta: "EmissionsData") -> None:
+        # Check if the run has been shutdown, if so do nothing
+        if (
+            self._simvue_run._shutdown_event
+            and self._simvue_run._shutdown_event.is_set()
+        ):
+            return
+
+        if self._meta_update:
```
[the remark this answers is not in the document]
Comment (kzscisoft) [reply simvue/eco.py:24]: Yes I missed `self._meta_update = False`

File: simvue/eco.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    def live_out(self, total: "EmissionsData", delta: "EmissionsData") -> None:
+        self.out(total, delta)
+
+
+class SimvueEmissionsTracker(EmissionsTracker):
+    def __init__(self, project_name: str, simvue_run: "Run") -> None:
+        self._simvue_run = simvue_run
+        super().__init__(
+            project_name=project_name,
+            measure_power_secs=0.1,
```
[the remark this answers is not in the document]
Comment (kzscisoft) [reply simvue/eco.py:51]: The high abundance of API calls is handled by the dispatcher but we can add this as an option and I think the default should be the same as resource metrics interval which is the heartbeat rate of 60s

[the remark this answers is not in the document]
Comment (alahiff) [reply simvue/eco.py:51]: Is this really collecting emissions data every 0.1 secs? What is the value of collecting it so often? I assume this must be a mistake?

File: simvue/eco.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            experiment_id=None,
+            experiment_name=None,
+            logging_logger=CodeCarbonOutput(simvue_run),
+            save_to_logger=True,
+            allow_multiple_runs=False,
```
[the remark this answers is not in the document]
Comment (kzscisoft) [reply simvue/eco.py:56]: Done

**Comment by alahiff** [comment]
The names of the metrics are possibly not compatible with the standard way "magic" metrics grouping works in the UI. By default the following will be grouped:
* codecarbon.emissions.delta, codecarbon.emissions.total
* codecarbon.energy_consumed.delta, codecarbon.energy_consumed.total

Does it make sense for deltas and totals to be displayed together?

**Comment by alahiff** [comment]
Is it possible to suppress this message to stdout?
```
[codecarbon WARNING @ 05:39:03] Multiple instances of codecarbon are allowed to run at the same time.
```
I also get this warning now when running the client:
```
/home/alahiff/.cache/pypoetry/virtualenvs/simvue-0vzJveOR-py3.10/lib/python3.10/site-packages/codecarbon/output_methods/file.py:52: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  df = pd.concat([df, pd.DataFrame.from_records([dict(total.values)])])
```

**Comment by alahiff** [comment]
When running this for 180 seconds I only get a single set of codecarbon metrics at the very end. Is this expected?

**Comment by AbyAbraham21** [comment]
"Keep emissions data off by default" - Why? I would assume you want it on by default?

**Comment by kzscisoft** [comment]
@AbyAbraham21 Due to the fact that setup is required on the host machine for the CC metrics to be valid and correct we should make the user accountable for their recording as opposed to it being our responsibility that their setup is correct. @alahiff and I agreed having these off by default made sense.

We will of course need to document enabling the feature.

**Pull request merged**